Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE for wtf/StringExtras

iidmsa opened the pull request:
#### 1e9d692bf39e99e20fb82691680a657a38d1a11e
<pre>
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE for wtf/StringExtras
<a href="https://bugs.webkit.org/show_bug.cgi?id=300067">https://bugs.webkit.org/show_bug.cgi?id=300067</a>
<a href="https://rdar.apple.com/problem/161862305">rdar://problem/161862305</a>

Reviewed by Chris Dumez.

Tests: Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp
       Tools/TestWebKitAPI/Tests/WTF/IntegerToStringConversion.cpp
       Tools/TestWebKitAPI/Tests/WebCore/FileMonitor.cpp
       Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::memmem):
* Source/WTF/wtf/StringExtras.h: Removed.
* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
* Source/WebCore/Modules/websockets/WebSocketHandshake.cpp:
* Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
* Source/WebCore/platform/win/ClipboardUtilitiesWin.cpp:
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
* Tools/TestWebKitAPI/Tests/WTF/IntegerToStringConversion.cpp:
* Tools/TestWebKitAPI/Tests/WebCore/FileMonitor.cpp:
* Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp:
* Tools/WebKitTestRunner/cairo/TestInvocationCairo.cpp:

Canonical link: <a href="https://commits.webkit.org/301029@main">https://commits.webkit.org/301029@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5244d63ca35a32eaea58c0d291cf478f7eaeeacb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124662 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44334 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35068 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131502 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76591 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126539 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45038 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52904 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94843 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62902 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127616 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35914 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111486 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75413 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34843 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29641 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74983 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116767 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105671 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29872 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134169 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123182 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51510 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39326 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103318 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51922 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103093 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26256 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48465 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26729 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48466 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51384 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57181 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/156292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50778 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/156292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54133 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52472 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -925,6 +924,32 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+
+    for (size_t i = 0; i <= lastPossiblePosition; ++i) {
+        auto candidateSpan = haystackSpan.subspan(i, needleLength);
+        if (std::equal(candidateSpan.begin(), candidateSpan.end(), needleSpan.begin()))
```
Can we rely on `equalSpans()` instead of `std::equal()` + iterators?

cdumez reviewed (approved):
LGTM with a suggestion.

cdumez reviewed (approved):
LGTM

iidmsa committed:
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE for wtf/StringExtras
https://bugs.webkit.org/show_bug.cgi?id=300067
rdar://problem/161862305

Reviewed by Chris Dumez.

Tests: Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp
       Tools/TestWebKitAPI/Tests/WTF/IntegerToStringConversion.cpp
       Tools/TestWebKitAPI/Tests/WebCore/FileMonitor.cpp
       Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::memmem):
* Source/WTF/wtf/StringExtras.h: Removed.
* Source/WebCore/Modules/fetch/FetchBodyConsumer.cpp:
* Source/WebCore/Modules/websockets/WebSocketHandshake.cpp:
* Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:
* Source/WebCore/platform/win/ClipboardUtilitiesWin.cpp:
* Tools/TestWebKitAPI/Tests/WTF/FileSystem.cpp:
* Tools/TestWebKitAPI/Tests/WTF/IntegerToStringConversion.cpp:
* Tools/TestWebKitAPI/Tests/WebCore/FileMonitor.cpp:
* Tools/TestWebKitAPI/Tests/WebCore/SharedBuffer.cpp:
* Tools/WebKitTestRunner/cairo/TestInvocationCairo.cpp:

Canonical link: https://commits.webkit.org/301029@main

shallawa commented on Source/WTF/wtf/StdLibExtras.h:
```diff
@@ -1018,6 +1017,32 @@ bool equalSpans(std::span<T, TExtent> a, std::span<U, UExtent> b)
+    if (haystackLength < needleLength)
+        return nullptr;
+
+    auto haystackSpan = unsafeMakeSpan(static_cast<const uint8_t*>(haystack), haystackLength);
+    auto needleSpan = unsafeMakeSpan(static_cast<const uint8_t*>(needle), needleLength);
```
This change is weird. `memmem()` is called only from two functions 
```
size_t find(std::span<T, TExtent> haystack, std::span<U, UExtent> needle)
size_t contains(std::span<T, TExtent> haystack, std::span<U, UExtent> needle)
```
Both of them take `std::spans`. To call `memmem()` with the current signature, they have to unpack their `std::spans` to `const void*` and `size_t`. Then the new `memmem()` packs them pack again into `std::spans`.

I would suggest moving this code to `find()` and make `contains()` call `find()`.

cdumez replied:
@shallawa How about these calls?
```
Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:    auto matchedStartPtr = static_cast<const uint8_t*>(memmem(contentStartPtr, contentLength, boundaryStartPtr, boundaryLength));
Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:    if ((end = static_cast<const uint8_t*>(memmem(contentStartPtr, contentLength, "\r\n\r\n", 4))))
Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp:    else if ((end = static_cast<const uint8_t*>(memmem(contentStartPtr, contentLength, "\n\n", 2))))
```

cdumez replied:
@shallawa Note that this `memmem()` implementation is only for the Windows port since other ports can just call the standard one.

shallawa replied:
I did not see these calls in CurlMultipartHandle.cpp because they are only included in the Windows and PlayStation cmakes. What I meant was to delete this customized `memmem()` implementation and to do something like this:
```
template<typename T, std::size_t TExtent, typename U, std::size_t UExtent>
size_t find(std::span<T, TExtent> haystack, std::span<U, UExtent> needle)
{
#if !HAVE(MEMMEM)
	// Use the above customized sub-span search. No need to call unsafeMakeSpan().
#else
	auto* result = memmem(...);
    if (!result)
        return notFound;
    return result - haystack.data();
#endif
}
#end

template<typename T, std::size_t TExtent, typename U, std::size_t UExtent>
size_t contains(std::span<T, TExtent> haystack, std::span<U, UExtent> needle)
{
	return find(...) != notFound;
}
```
And for the calls in CurlMultipartHandle.cpp they can use `find()` instead since they all search in a Vector.

iidmsa replied:
@shallawa Suggestion addressed here: https://github.com/WebKit/WebKit/pull/52700
